builder/cloudstack/step_create_instance.go

```diff
@@ -65,7 +115,29 @@ func (s *stepCreateInstance) Run(state multistep.StateBag) multistep.StepAction
 	}
 
 	if config.UserData != "" {
-		ud, err := getUserData(config.UserData, config.HTTPGetOnly)
+		var httpPort uint = state.Get("http_port").(uint)
+		hostIp, err := hostIP()
+		if err != nil {
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+		common.SetHTTPIP(hostIp)
+
+		s.Ctx.Data = &createInstanceTemplateData{
+			hostIp,
+			httpPort,
+		}
+
+		renderedUserData, err := interpolate.Render(config.UserData, &s.Ctx)
+		if err != nil {
+			err := fmt.Errorf("Error rendering user_data: %s", err)
+			state.Put("error", err)
+			ui.Error(err.Error())
+			return multistep.ActionHalt
+		}
+
+		ud, err := getUserData(renderedUserData, config.HTTPGetOnly)
```

Comment:
I'm not entirely sure what you mean here, @svanharmelen. If you mean that we put this:

```
                renderedUserData, err := interpolate.Render(config.UserData, &s.Ctx)
 		if err != nil {
 			err := fmt.Errorf("Error rendering user_data: %s", err)
 			state.Put("error", err)
 			ui.Error(err.Error())
 			return multistep.ActionHalt
 		}
```
in its own function named  ```generateUserData```, then I'm not sure I think it worth it. But maybe you mean something else?